scipy/special/tests/test_basic.py

```diff
@@ -3658,3 +3659,56 @@ def test_runtime_warning():
     with pytest.warns(RuntimeWarning,
                       match=r'Too many predicted coefficients'):
         mathieu_even_coef(1000, 1000)
+
+
+class TestStirling2:
+    table = [
+        [1],
+        [0,1],
+        [0,1,1],
+        [0,1,3,1],
+        [0,1,7,6,1],
+        [0,1,15,25,10,1],
+        [0,1,31,90,65,15,1],
+        [0,1,63,301,350,140,21,1],
+        [0,1,127,966,1701,1050,266,28,1],
+        [0,1,255,3025,7770,6951,2646,462,36,1],
+        [0,1,511,9330,34105,42525,22827,5880,750,45,1],
+    ]
+
+    def test_table_cases(self):
+        for n in range(len(self.table)):
+            for k in range(len(self.table[n])):
+                assert_equal(self.table[n][k], stirling2(n,k))
+
+    def test_valid_single_integer(self):
+        assert_equal(stirling2(0, 0), self.table[0][0])
+        assert_equal(stirling2(4, 2), self.table[4][2])
+
+    def test_negative_integer(self):
+        assert_equal(stirling2(-1, -1), 0)
+        assert_equal(stirling2(-1, 2), 0)
+        assert_equal(stirling2(2, -1), 0)
```

Comment:
Negative Stirling number should not be zero:
https://wikipedia.org/wiki/Stirling_number#Stirling_numbers_with_negative_integral_values

So unless you feel like implementing this as well, I think that throwing an exception in this case would be best.

Reply:
Fun fact: `stirling2(-n, -k) == stirling1(k, n)`

Reply:
> Fun fact: `stirling2(-n, -k) == stirling1(k, n)`. So `stirling2(-1, -1)` should actually return `1`.

This is a choice of implementation, there are subtleties to using negative numbers-e.g. there are 2 conventions as indicated in the wikipedia article and the calculations for both the first and second kind Stirling numbers seem different enough that we can't put that into this same framework. 

> Negative Stirling number should not be zero: https://wikipedia.org/wiki/Stirling_number#Stirling_numbers_with_negative_integral_values
> 

and 

> So unless you feel like implementing this, I think that throwing an exception in this case would be best.

The zero behavior is a relic from when we had these implemented as `ufunc`s in an earlier commit on this PR. 

The problem with throwing an exception here is many of the calculations are perfectly fine and throwing an exception b/c of a (possibly) single erroneous input wastes all the calculation. The usual approach in `ufuncs` is then something like a `Nan` for the specific input-I work on a switch `Nan`-thanks for pointing this out.

Reply:
Hmm, now I'm thinking a bit more about this, if we stick with the interpretation in the docs, (K non-empty subsets of N distinct items, then a negative number doesn't make sense here). While I'm not disagreeing with the mathematical validity of the claim I'm not sure we'd want to surprise users in this way by having `Nan`. @steppi do you have any thoughts here on whether we should put the negative input values as `Nan` instead of `0`. The docs currently state `0` so the behavior-as currently implemented in the pr-seems valid but the question is whether a `Nan` might make more sense?

Reply:
I believe that similarly `scipy.special.binom` also returns `nan`'s for negative input, so I agree that similar behaviour makes most sense 👌🏻

Reply:
> I believe that similarly `scipy.special.binom` also returns `nan`'s for negative input, so I agree that similar behaviour makes most sense 👌🏻

IIUC the plan is that the `binom` is getting swapped w/`comb` and we're aiming to match the convention on `comb`  method which returns `0` for these types of inputs. The duality between +/- would be handled easily by a caller once we have `stirling1` implemented-that's a planned follow up PR.

Reply:
> @steppi do you have any thoughts here on whether we should put the negative input values as `Nan` instead of `0`?

If we don't implement an extension to negative values now, I think we should follow the convention of `comb` and return  `0`.  It's the least surprising thing in my opinion and also I'm a little uneasy about mixing integer and float return types.

Reply:
> > @steppi do you have any thoughts here on whether we should put the negative input values as `Nan` instead of `0`?
> 
> If we don't implement an extension to negative values now, I think we should follow the convention of `comb` and return `0`. It's the least surprising thing in my opinion and also I'm a little uneasy about mixing integer and float return types.

Ah yeah, good point. I hadn't considered that then we'd be mixing types too. That would basically break everything that's working nicely now with the arbitrary int precision. I agree let's keep it 0.

IIUC The negative values would be handled via `stirling1` so the caller can always do a branching logic. If there is enough demand for a generic single function that handles both positive and negative, we can always add a wrapping function after we have both stirling variants in place but my preference for now is to keep them separate.